### ISSUE TITLE
STK:  Build error on OS X + LLVM/Clang

### ISSUE THREAD

**Issue by sayerhs** [trilinos/Trilinos#6012]
## Bug Report
@trilinos/stk

I am getting a compilation error with VOTD trilinos/develop branch on OS X + Clang 9 + OpenMPI 3.1.1. 

```
[34/382] Building CXX object packages/stk/stk_mesh/stk_mesh/base/CMakeFiles/stk_mesh_base.dir/BulkData.cpp.o
FAILED: packages/stk/stk_mesh/stk_mesh/base/CMakeFiles/stk_mesh_base.dir/BulkData.cpp.o 
/usr/local/bin/mpicxx   -I. -Ipackages/stk/stk_mesh/stk_mesh/base -I../packages/stk/stk_mesh -I../packages/stk/stk_mesh/stk_mesh/base -I../packages/stk/stk_topology -Ipackages/stk/stk_util/stk_util/diag -I../packages/stk/stk_util/stk_util/diag -I../packages/stk/stk_util/stk_util/environment -Ipackages/stk/stk_util/stk_util/command_line -I../packages/stk/stk_util -Ipackages/stk/stk_util -I../packages/stk/stk_util/stk_util/command_line -Ipackages/stk/stk_util/stk_util/registry -I../packages/stk/stk_util/stk_util/registry -Ipackages/stk/stk_util/stk_util/parallel -I../packages/stk/stk_util/stk_util/parallel -Ipackages/stk/stk_util/stk_util/util -I../packages/stk/stk_util/stk_util/util -Ipackages/kokkos/core/src -I../packages/kokkos/core/src -I/usr/local/Cellar/boost/1.70.0/include -Ipackages/shards/src -I../packages/shards/src -std=c++11 -O3 -DNDEBUG -MD -MT packages/stk/stk_mesh/stk_mesh/base/CMakeFiles/stk_mesh_base.dir/BulkData.cpp.o -MF packages/stk/stk_mesh/stk_mesh/base/CMakeFiles/stk_mesh_base.dir/BulkData.cpp.o.d -o packages/stk/stk_mesh/stk_mesh/base/CMakeFiles/stk_mesh_base.dir/BulkData.cpp.o -c ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:238:8: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "From\\To:  NODE     EDGE     FACE     ELEM\n";
   ~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [43]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[43]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[43]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:239:8: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "NODE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[0][i] << "  "; } out << "\n";
   ~~~ ^  ~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [11]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[11]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[11]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:239:57: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'int')
   out << "NODE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[0][i] << "  "; } out << "\n";
                                                    ~~~ ^  ~~~~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const value_type' (aka 'const stk::mesh::ConnectivityType') to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'const stk::mesh::ConnectivityType'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'const stk::mesh::ConnectivityType'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:239:91: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "NODE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[0][i] << "  "; } out << "\n";
                                                                                      ~~~ ^  ~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [2]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[2]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[2]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:240:8: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "EDGE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[1][i] << "  "; } out << "\n";
   ~~~ ^  ~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [11]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[11]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[11]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:240:57: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'int')
   out << "EDGE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[1][i] << "  "; } out << "\n";
                                                    ~~~ ^  ~~~~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const value_type' (aka 'const stk::mesh::ConnectivityType') to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'const stk::mesh::ConnectivityType'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'const stk::mesh::ConnectivityType'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:240:91: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "EDGE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[1][i] << "  "; } out << "\n";
                                                                                      ~~~ ^  ~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [2]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[2]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[2]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:241:8: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "FACE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[2][i] << "  "; } out << "\n";
   ~~~ ^  ~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [11]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[11]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[11]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:241:57: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'int')
   out << "FACE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[2][i] << "  "; } out << "\n";
                                                    ~~~ ^  ~~~~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const value_type' (aka 'const stk::mesh::ConnectivityType') to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'const stk::mesh::ConnectivityType'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'const stk::mesh::ConnectivityType'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:241:91: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "FACE   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[2][i] << "  "; } out << "\n";
                                                                                      ~~~ ^  ~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [2]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[2]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[2]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:242:8: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "ELEM   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[3][i] << "  "; } out << "\n";
   ~~~ ^  ~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [11]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[11]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[11]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:242:57: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'int')
   out << "ELEM   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[3][i] << "  "; } out << "\n";
                                                    ~~~ ^  ~~~~~~~~~~~~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const value_type' (aka 'const stk::mesh::ConnectivityType') to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'const stk::mesh::ConnectivityType'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'const stk::mesh::ConnectivityType'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
In file included from ../packages/stk/stk_mesh/stk_mesh/base/BulkData.cpp:35:
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:242:91: error: invalid operands to binary expression ('std::ostream' (aka 'basic_ostream<char>') and 'const char *')
   out << "ELEM   :  "; for (int i=0 ; i<4 ; ++i) { out << map.m_map[3][i] << "  "; } out << "\n";
                                                                                      ~~~ ^  ~~~~
../packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp:236:23: note: candidate function not viable: no known conversion from 'const char [2]' to 'const stk::mesh::ConnectivityMap' for 2nd argument
inline std::ostream & operator<<(std::ostream &out, const ConnectivityMap & map)
                      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:5487:1: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-2>' against 'char const[2]'
operator<<(basic_ostream<_CharT, _Traits>& __os, shared_ptr<_Yp> const& __p);
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:3898:1: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2>' against 'char const[2]'
operator<<(basic_ostream<_CharT, _Traits>& __os,
^
```

### Steps to Reproduce

- Trilinos SHA: 39afaf5763e7c48f992a039d34e51e231e3b3f0f
- Trilinos configure [command line](https://github.com/sayerhs/exawind-builder/blob/master/codes/trilinos.bash#L63)
- [cmake_output.log](https://github.com/trilinos/Trilinos/files/3678015/cmake_output.log)
- [make_output.log](https://github.com/trilinos/Trilinos/files/3678016/make_output.log)

**Comment by sayerhs** [comment]
Cc: @alanw0

**Comment by jhux2** [comment]
@trilinos/stk

**Comment by alanw0** [comment]
@sayerhs I'll take care of this as soon as possible.
Dave has found that when apple reports 'clang 9' it is really clang 4. On the sierra systems the oldest clang version we have is 5, so I may have trouble reproducing this, but I think I can see what needs to be done anyway.

**Comment by sayerhs** [comment]
@alanw0 Does this help:

```
$ clang --version
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by sayerhs** [comment]
@alanw0 I was able to get it compiled with this patch

```diff
diff --git a/packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp b/packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp
index 7c1c9fbd7f..a0c516510c 100644
--- a/packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp
+++ b/packages/stk/stk_mesh/stk_mesh/base/ConnectivityMap.hpp
@@ -37,6 +37,7 @@

 #include <stk_mesh/base/Types.hpp>
 #include <array>
+#include <iostream>

 namespace stk  {
 namespace mesh {
```

**Comment by alanw0** [comment]
@sayerhs ok that's good, that will keep you working until I get a fix pushed up to trilinos.

**Comment by alanw0** [comment]
This was fixed by PR #6013
